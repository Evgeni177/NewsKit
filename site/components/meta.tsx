import * as React from 'react';

const Meta = (): JSX.Element => (
  <React.Fragment>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
      key="viewport"
    />
    <link rel="icon" type="image/x-icon" href="/static/favicon.ico" />
    <link
      rel="apple-touch-icon"
      sizes="180x180"
      href="/static/apple-touch-icon.png"
    />
    <link
      rel="icon"
      type="image/png"
      sizes="16x16"
      href="/static/favicon-16x16.png"
    />
    <link
      rel="icon"
      type="image/png"
      sizes="32x32"
      href="/static/favicon-32x32.png"
    />
    <link
      rel="icon"
      type="image/png"
      sizes="192x192"
      href="/static/android-chrome-192x192.png"
    />
    <link rel="manifest" href="/static/site.webmanifest" />
    <meta name="msapplication-TileColor" content="#ffffff" />
    <meta name="theme-color" content="#ffffff" />
  </React.Fragment>
);

export default Meta;
