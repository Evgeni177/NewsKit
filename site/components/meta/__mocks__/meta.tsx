import React from 'react';

export const Meta: React.FC = props => (
  <div>
    Meta Component with props:
    {JSON.stringify(props, null, 2)}
  </div>
);
