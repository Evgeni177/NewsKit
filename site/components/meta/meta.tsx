import React from 'react';
import {
  Stack,
  Flag,
  Block,
  Divider,
  Button,
  StackChild,
  AlignSelfValues,
  Link,
  IconFilledGitHub,
  Flow,
  Headline,
  FlagSize,
  IconFilledFigma,
  Hidden,
  Cell,
} from 'newskit';
import {MetaProps, MetaFlagStylePresets} from './types';

export const Meta = ({status, introduced, codeUrl, figmaUrl}: MetaProps) => (
  <Cell xs={12} md={10} lg={8} mdOffset={1}>
    <Block spaceStack="space110">
      <Block spaceStack="space050">
        <Hidden xs sm md lg>
          <Stack flow={Flow.HorizontalCenter} stackDistribution="space-between">
            <Stack flow={Flow.HorizontalTop} spaceInline="space060">
              <Stack flow={Flow.VerticalLeft} spaceInline="space030">
                <Headline overrides={{typographyPreset: 'utilityLabel020'}}>
                  Status
                </Headline>
                <Flag
                  size={FlagSize.Large}
                  overrides={{
                    stylePreset: MetaFlagStylePresets[status],
                    typographyPreset: 'utilityLabel020',
                  }}
                >
                  {status}
                </Flag>
              </Stack>
              <StackChild alignSelf={AlignSelfValues.Stretch}>
                <Divider vertical />
              </StackChild>
              <Stack flow={Flow.VerticalLeft} spaceInline="space050">
                <Headline overrides={{typographyPreset: 'utilityLabel020'}}>
                  Introduced
                </Headline>
                <Link
                  href="http://localhost:8081"
                  overrides={{typographyPreset: 'utilityBody030'}}
                >
                  {introduced}
                </Link>
              </Stack>
            </Stack>
            <Stack flow="horizontal-center" spaceInline="space040">
              <Button
                overrides={{stylePreset: 'buttonOutlinedSecondary'}}
                href={codeUrl}
              >
                <IconFilledGitHub />
                View source code
              </Button>
              {figmaUrl && (
                <Button
                  overrides={{
                    stylePreset: 'buttonOutlinedSecondary',
                    minWidth: '176px',
                  }}
                  href={figmaUrl}
                >
                  <IconFilledFigma />
                  View Design
                </Button>
              )}
            </Stack>
          </Stack>
        </Hidden>
        <Hidden xl>
          <Stack
            flow={Flow.VerticalLeft}
            stackDistribution="space-between"
            spaceInline="space050"
          >
            <Stack flow={Flow.HorizontalTop} spaceInline="space060">
              <Stack flow={Flow.VerticalLeft} spaceInline="space030">
                <Headline overrides={{typographyPreset: 'utilityLabel020'}}>
                  Status
                </Headline>
                <Flag
                  size={FlagSize.Large}
                  overrides={{
                    stylePreset: MetaFlagStylePresets[status],
                    typographyPreset: 'utilityLabel020',
                  }}
                >
                  {status}
                </Flag>
              </Stack>
              <StackChild alignSelf={AlignSelfValues.Stretch}>
                <Divider vertical />
              </StackChild>
              <Stack flow={Flow.VerticalLeft} spaceInline="space050">
                <Headline overrides={{typographyPreset: 'utilityLabel020'}}>
                  Introduced
                </Headline>
                <Link
                  href="http://localhost:8081"
                  overrides={{typographyPreset: 'utilityBody030'}}
                >
                  {introduced}
                </Link>
              </Stack>
            </Stack>
            <Stack flow="horizontal-center" spaceInline="space040">
              <Button overrides={{stylePreset: 'buttonOutlinedSecondary'}}>
                <IconFilledGitHub />
                View source code
              </Button>
              {figmaUrl && (
                <Button
                  overrides={{
                    stylePreset: 'buttonOutlinedSecondary',
                    minWidth: '176px',
                  }}
                  href={figmaUrl}
                >
                  <IconFilledFigma />
                  View Design
                </Button>
              )}
            </Stack>
          </Stack>
        </Hidden>
      </Block>
      <Divider />
    </Block>
  </Cell>
);
