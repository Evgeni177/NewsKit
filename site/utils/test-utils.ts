import {
  renderToFragmentWithThemeFactory,
  renderWithThemeFactory,
} from 'newskit/test/test-utils';
import {docsThemeLight} from '../theme/doc-theme';

export const renderWithTheme = renderWithThemeFactory(docsThemeLight);

export const renderToFragmentWithTheme = renderToFragmentWithThemeFactory(
  docsThemeLight,
);
