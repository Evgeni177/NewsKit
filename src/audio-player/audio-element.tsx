/* eslint-disable jsx-a11y/media-has-caption */
import React, {
  useRef,
  useImperativeHandle,
  forwardRef,
  RefObject,
  Ref,
} from 'react';

const getValueInRange = (value: number, [start, end]: [number, number]) =>
  Math.max(start, Math.min(end, value));

export interface AudioHandler {
  play: () => void;
  pause: () => void;
  setVolume: (newVolume: number) => number;
  setCurrentTime: (newTime: number) => number;
}

export const useAudioHandler = (
  localRef: RefObject<HTMLAudioElement>,
  parentRef: Ref<AudioHandler>,
) => {
  useImperativeHandle(parentRef, () => ({
    play: () => {
      const playerNode = localRef.current;
      if (playerNode) {
        playerNode.play();
      }
    },
    pause: () => {
      const playerNode = localRef.current;
      if (playerNode) {
        playerNode.pause();
      }
    },
    setCurrentTime: newTime => {
      const playerNode = localRef.current!;
      const timeRange = playerNode.seekable;
      if (timeRange.length > 0) {
        const value = getValueInRange(newTime, [
          timeRange.start(0),
          timeRange.end(timeRange.length - 1),
        ]);
        playerNode.currentTime = value;
        return value;
      }
      return playerNode.currentTime;
    },
    setVolume: newVolume => {
      const playerNode = localRef.current!;
      const value = getValueInRange(newVolume, [0, 1]);
      playerNode.volume = value;
      return value;
    },
  }));
};

export interface AudioElementProps
  extends React.AudioHTMLAttributes<HTMLAudioElement> {
  captionSrc?: string;
}

export const AudioElement = forwardRef<AudioHandler, AudioElementProps>(
  ({captionSrc, ...props}, parentRef) => {
    const localRef = useRef<HTMLAudioElement>(null);
    useAudioHandler(localRef, parentRef);
    return (
      <audio ref={localRef} {...props}>
        {captionSrc && <track kind="captions" src={captionSrc} />}
      </audio>
    );
  },
);
