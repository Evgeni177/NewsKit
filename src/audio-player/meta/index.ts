export * from './audio-element';
export * from './player-meta';
