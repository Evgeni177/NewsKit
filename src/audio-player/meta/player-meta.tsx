import React from 'react';
import {
  ProgrammeTime,
  ProgrammeTitle,
  ProgrammeDescription,
  ProgrammeTags,
  ImageContainer,
  CenterProp,
} from './styled';
import {Grid, Cell} from '../../grid';
import {Image} from '../../image';
import {Stack, Flow, StackDistribution} from '../../stack';
import {Block} from '../../block';
import {Flag} from '../../flag';

interface DescriptionMetaProps {
  title?: string;
  time?: string;
  description?: string;
  live?: boolean;
  flag?: React.ComponentType | string;
  tags?: string[];
}

const renderFlag = (CustomFlag: React.ComponentType | string) =>
  typeof CustomFlag === 'string' ? (
    <Flag $stylePreset="flagSolidLive">{CustomFlag}</Flag>
  ) : (
    <CustomFlag />
  );

const DescriptionsMeta: React.FC<DescriptionMetaProps & CenterProp> = ({
  center = false,
  time,
  flag,
  title,
  description,
  tags,
}) => (
  <Stack
    flow={center ? Flow.VerticalCenter : Flow.VerticalLeft}
    stackDistribution={StackDistribution.Start}
  >
    {(flag || time) && (
      <Block $margin="spaceStack050">
        <Stack
          flow={center ? Flow.VerticalCenter : Flow.HorizontalCenter}
          stackDistribution={
            center ? StackDistribution.SpaceBetween : StackDistribution.Start
          }
          space={center ? 'sizing020' : 'sizing040'}
        >
          {flag && renderFlag(flag)}
          {time && <ProgrammeTime>{time}</ProgrammeTime>}
        </Stack>
      </Block>
    )}
    {title && (
      <Block $margin="spaceStack040">
        <ProgrammeTitle center={center}>{title}</ProgrammeTitle>
      </Block>
    )}
    {description && (
      <ProgrammeDescription center={center}>{description}</ProgrammeDescription>
    )}
    {tags && tags.length > 0 && (
      <ProgrammeTags>
        {tags.map((tag, i) => `${tag}${i <= tags.length - 2 ? ' | ' : ''}`)}
      </ProgrammeTags>
    )}
  </Stack>
);

interface PlayerImageProps {
  imgSrc: string;
  imgAlt: string;
}

const PlayerImage: React.FC<PlayerImageProps> = ({imgSrc, imgAlt}) => (
  <ImageContainer>
    <Image
      src={imgSrc}
      alt={imgAlt}
      aspectHeight="1"
      aspectWidth="1"
      $stylePreset="maskRound010"
    />
  </ImageContainer>
);

export type PlayerMetaProps = PlayerImageProps & DescriptionMetaProps;

export const PlayerMeta: React.FC<PlayerMetaProps> = React.memo(
  ({imgSrc, imgAlt, ...props}) => (
    <Grid>
      <Cell md={5} xsHidden smHidden>
        <Stack
          flow={Flow.HorizontalTop}
          stackDistribution={StackDistribution.End}
        >
          <PlayerImage imgSrc={imgSrc} imgAlt={imgAlt} />
        </Stack>
      </Cell>
      <Cell md={7} xsHidden smHidden>
        <DescriptionsMeta {...props} />
      </Cell>
      <Cell mdHidden lgHidden xs={12}>
        <Stack
          flow={Flow.HorizontalCenter}
          stackDistribution={StackDistribution.Center}
        >
          <PlayerImage imgSrc={imgSrc} imgAlt={imgAlt} />
        </Stack>
      </Cell>
      <Cell mdHidden lgHidden xs={12}>
        <DescriptionsMeta {...props} center />
      </Cell>
    </Grid>
  ),
);
