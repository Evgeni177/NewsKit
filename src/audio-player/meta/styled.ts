import {
  styled,
  getColorFromTheme,
  getSizingFromTheme,
  getTypePresetFromTheme,
  getFontsFromTheme,
  getFontSizingFromTheme,
} from '../../utils/style';
import {H1} from '../../typography';
import {Tag} from '../../tag';

export const LiveTag = styled(Tag)`
  width: ${getSizingFromTheme('sizing080')};
  height: ${getSizingFromTheme('sizing050')};
  padding: 0;
  text-align: center;
  background-color: ${getColorFromTheme('semanticNegative010')};
  color: ${getColorFromTheme('inkInverse')};
  text-transform: uppercase;
  margin-right: ${getSizingFromTheme('sizing040')};
`;

export const Label = styled.span`
  color: ${getColorFromTheme('inkSubtle')};
  font-weight: ${getFontsFromTheme('fontWeight010')};
`;

export const ProgrammeTime = styled(Label)`
  ${getTypePresetFromTheme('meta010')};
  display: inline-block;
`;

export const ProgrammeTitle = styled(H1)`
  ${getTypePresetFromTheme('heading040')};
  margin: 0;
`;

export const ProgrammeDescription = styled(Label)`
  ${getTypePresetFromTheme('subhead010')};
`;

export const ProgrammeTags = styled(Label)`
  ${getTypePresetFromTheme('meta010')};
  ${getFontSizingFromTheme('fontSize020', 'fontLineHeight040')};
`;

export const ImageContainer = styled.div`
  max-width: 212px;
  max-height: 212px;
  margin-left: auto;
`;
