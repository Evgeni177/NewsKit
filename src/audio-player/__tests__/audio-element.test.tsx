/* eslint-disable no-unused-expressions */
import {AudioElement, useAudioHandler} from '../meta/audio-element';
import {renderToFragmentWithTheme, renderHook} from '../../test/test-utils';

describe('AudioElement', () => {
  describe('renders correctly', () => {
    test('default <audio>', () => {
      const fragment = renderToFragmentWithTheme(AudioElement);
      expect(fragment).toMatchSnapshot();
    });

    test('<audio> with caption', () => {
      const fragment = renderToFragmentWithTheme(AudioElement, {
        src: 'https://radio.talkradio.co.uk/stream',
        captionSrc: 'captions.vtt',
      });
      expect(fragment).toMatchSnapshot();
    });
  });

  describe('useAudioHandler', () => {
    describe('togglePlay', () => {
      test('calls play method on player, if isPlaying is false', () => {
        const play = jest.fn();
        const pause = jest.fn();
        const playerRef = {
          current: {
            play,
            pause,
          },
        } as any;
        const propRef = {
          current: {
            togglePlay: jest.fn(),
          },
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));
        const isPlaying = propRef.current.togglePlay(true);
        expect(isPlaying).toBe(false);
        expect(playerRef.current.pause).toHaveBeenCalled;
        expect(playerRef.current.play).not.toHaveBeenCalled;
      });

      test('calls pause method on player, if isPlaying is true', () => {
        const play = jest.fn();
        const pause = jest.fn();
        const playerRef = {
          current: {
            play,
            pause,
          },
        } as any;
        const propRef = {
          current: {
            togglePlay: jest.fn(),
          },
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));
        const isPlaying = propRef.current.togglePlay(false);
        expect(isPlaying).toBe(true);
        expect(playerRef.current.pause).not.toHaveBeenCalled;
        expect(playerRef.current.play).toHaveBeenCalled;
      });
    });

    describe('setCurrentTime', () => {
      test('sets passed currentTime value onto the player', () => {
        const playerRef = {
          current: {
            currentTime: 0,
            seekable: {
              start: jest.fn().mockReturnValue(0),
              end: jest.fn().mockReturnValue(100),
              length: 1,
            },
          },
        } as any;
        const propRef = {
          current: {
            setCurrentTime: jest.fn(),
          },
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));
        const currentTime = propRef.current.setCurrentTime(50);
        expect(currentTime).toEqual(50);
        expect(playerRef.current.currentTime).toEqual(50);
        expect(playerRef.current.seekable.start).toHaveBeenCalledWith(0);
        expect(playerRef.current.seekable.end).toHaveBeenCalledWith(0);
      });
      test('does not set passed currentTime if timerange length is 0', () => {
        const playerRef = {
          current: {
            currentTime: 0,
            seekable: {
              start: jest.fn().mockReturnValue(0),
              end: jest.fn().mockReturnValue(100),
              length: 0,
            },
          },
        } as any;
        const propRef = {
          current: {
            setCurrentTime: jest.fn(),
          },
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));
        const currentTime = propRef.current.setCurrentTime(50);
        expect(currentTime).toEqual(0);
        expect(playerRef.current.currentTime).toEqual(0);
        expect(playerRef.current.seekable.start).not.toHaveBeenCalled();
        expect(playerRef.current.seekable.end).not.toHaveBeenCalled();
      });

      test('sets the minimum timerange value when passed currentTime is smaller minimum', () => {
        const playerRef = {
          current: {
            currentTime: 0,
            seekable: {
              start: jest.fn().mockReturnValue(10),
              end: jest.fn().mockReturnValue(20),
              length: 2,
            },
          },
        } as any;
        const propRef = {
          current: {
            setCurrentTime: jest.fn(),
          },
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));
        const currentTime = propRef.current.setCurrentTime(9);
        expect(currentTime).toEqual(10);
        expect(playerRef.current.currentTime).toEqual(10);
        expect(playerRef.current.seekable.start).toHaveBeenCalledWith(0);
        expect(playerRef.current.seekable.end).toHaveBeenCalledWith(1);
      });

      test('sets the maximum timerange value when passed currentTime is smaller maximum', () => {
        const playerRef = {
          current: {
            currentTime: 0,
            seekable: {
              start: jest.fn().mockReturnValue(10),
              end: jest.fn().mockReturnValue(20),
              length: 2,
            },
          },
        } as any;
        const propRef = {
          current: {
            setCurrentTime: jest.fn(),
          },
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));
        const currentTime = propRef.current.setCurrentTime(90);
        expect(currentTime).toEqual(20);
        expect(playerRef.current.currentTime).toEqual(20);
        expect(playerRef.current.seekable.start).toHaveBeenCalledWith(0);
        expect(playerRef.current.seekable.end).toHaveBeenCalledWith(1);
      });

      test('returns currentTime value if audio not seekable', () => {
        const playerRef = {
          current: {
            currentTime: 12345,
            seekable: {
              start: jest.fn(),
              end: jest.fn(),
              length: 0,
            },
          },
        } as any;
        const propRef = {
          current: {
            setCurrentTime: jest.fn(),
          },
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));
        const currentTime = propRef.current.setCurrentTime(50);
        expect(currentTime).toEqual(12345);
        expect(playerRef.current.currentTime).toEqual(12345);
        expect(playerRef.current.seekable.start).not.toHaveBeenCalled();
        expect(playerRef.current.seekable.end).not.toHaveBeenCalled();
      });
    });

    describe('setVolume', () => {
      test('sets volume to 0.5', () => {
        const playerRef = {
          current: {},
        } as any;
        const propRef = {
          current: {},
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));

        const newVolume = propRef.current.setVolume(0.5);

        expect(newVolume).toEqual(0.5);
        expect(playerRef.current.volume).toEqual(0.5);
      });

      test('bounds volume to a max of 1', () => {
        const playerRef = {
          current: {},
        } as any;
        const propRef = {
          current: {},
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));

        const newVolume = propRef.current.setVolume(11);

        expect(newVolume).toEqual(1);
        expect(playerRef.current.volume).toEqual(1);
      });

      test('bounds volume to a min of 0', () => {
        const playerRef = {
          current: {},
        } as any;
        const propRef = {
          current: {},
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));

        const newVolume = propRef.current.setVolume(-0.4);

        expect(newVolume).toEqual(0);
        expect(playerRef.current.volume).toEqual(0);
      });
    });

    describe('setVolume', () => {
      test('sets passed volume value onto the player', () => {
        const playerRef = {
          current: {
            volume: 1,
          },
        } as any;
        const propRef = {
          current: {
            setVolume: jest.fn(),
          },
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));
        const volume = propRef.current.setVolume(0.5);
        expect(volume).toEqual(0.5);
        expect(playerRef.current.volume).toEqual(0.5);
      });

      test('sets the minimum timerange value when passed volume is smaller minimum', () => {
        const playerRef = {
          current: {
            volume: 1,
          },
        } as any;
        const propRef = {
          current: {
            setVolume: jest.fn(),
          },
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));
        const volume = propRef.current.setVolume(-9);
        expect(volume).toEqual(0);
        expect(playerRef.current.volume).toEqual(0);
      });

      test('sets the maximum timerange value when passed volume is smaller maximum', () => {
        const playerRef = {
          current: {
            volume: 1,
          },
        } as any;
        const propRef = {
          current: {
            setVolume: jest.fn(),
          },
        } as any;
        renderHook(() => useAudioHandler(playerRef, propRef));
        const volume = propRef.current.setVolume(90);
        expect(volume).toEqual(1);
        expect(playerRef.current.volume).toEqual(1);
      });
    });
  });
});
