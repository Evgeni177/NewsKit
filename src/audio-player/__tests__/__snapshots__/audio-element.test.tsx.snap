// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioElement renders correctly with default props 1`] = `
<DocumentFragment>
  <audio
    autoplay=""
  />
</DocumentFragment>
`;

exports[`AudioElement renders with caption and src 1`] = `
<DocumentFragment>
  <audio
    autoplay=""
    src="https://radio.talkradio.co.uk/stream"
  >
    <track
      kind="captions"
      src="captions.vtt"
    />
  </audio>
</DocumentFragment>
`;
